manpages/xgvis.1
.TH XGVis 1 "20 November 1998"
.SH NAME
.B xgvis
\- interactive multidimensional scaling using xgobi for display
.SH SYNOPSIS
.B xgvis
[
.B \-dims
n
] [
.B \-stepsize
x
]
fname
.SH DESCRIPTION
.I xgvis
is an interactive multidimensional scaling (MDS) program that
consists of a control panel to manipulate the parameters of the
MDS stress function and an xgobi window for data display.  It
can be used either for visualization of dissimilarity data, for
dimension reduction, or for graph layout.  Graph layout is
usually done in 2D, but xgvis allows layouts in arbitrary
dimensions, 3D being the default.  It permits missing values,
which can be used to implement multidimensional unfolding.

It is used as follows:
.RS .25i
.IP o 2
Start up with a user-supplied or random starting configuration.
.IP o 2
Initiate MDS optimization.  As the optimization algorithm does
its work, observe both the changing stress values and the animated point
configuration.
.IP o 2
Examine the optimized configuration with 3D rotations and
grand tours and drill-down operations such as labeling and linked brushing.
.IP o 2
Enhance the configuration with glyphs, colors and lines for interpretation.
.IP o 2
Modifiy the dissimilarities, the stress function, and/or the 
configuration by
.RE
.RS .6i
.IP - 2
transforming the dissimilarities
.IP - 2
choosing weights or cutting extreme dissimilarities
.IP - 2
restricting MDS to subsets
.IP - 2
hand editing the existing configuration or
.IP - 2
selecting a new random configuration
.RE
.SH OPTIONS

.B \-dims
n
.LP
.RS .25i
Specify n, the dimensionality of the embedding space.
By default, its value is 3.
.RE

.B \-stepsize
x
.LP
.RS .25i
Specify the initial stepsize; its default value is 0.1.
.RE

.SH FILES
XGVis uses a set of input files with a common root (which we'll call
.B fname
in the man page) and a suffix.  First, each set must contain a file
with the suffix .edges, .dat or .dist.
.RE

.nf
.B  fname.dist
.fi
.RS .25i
Distance matrix: an n x n matrix, with entries separated by
white space and one row per line.
.RE

.nf
.B  fname.edges
.fi
.RS .25i
Line segments: specifications for the pattern of line segments which
connect pairs of points.  The file must contain at least two numbers per
line.  The first two numbers represent the row numbers of the two
points that should be connected.  (This is exactly like the structure
of a fname.lines file in xgobi.)  In addition, if a third number is
present, it is taken to be an edge weight.

If this line is present and the distance matrix is absent, then
the distance matrix is computed from it, with each
edge representing a distance of one.
.RE

.nf
.B fname.dat
.B fname.pos
.fi
.RS .25i
Starting positions:  an n x p matrix, with entries separated by
white space and one row per line. 

If this file is present and a distance file is absent, the
distance matrix is computed from it.
.RE

.nf
.B fname.lines
.fi
.RS .25i

Line segments: specifications for the pattern of line segments which
connect pairs of points.  If this file is present, then the edges
file is used to create the distance matrix but the lines file is
used to draw the edges.
.RE

Many additional files can be specified.  See the xgobi man
page for descriptions of 
.B fname.col
(variable labels),
.B fname.row
(case labels),
.B fname.colors
(glyph colors), and
.B fname.linecolors
(line colors),
in particular.

.RE
.SH AUTHORS
.nf
Andreas Buja (andreas@research.att.com)
Deborah Swayne (dfs@research.att.com)
.fi
.SH CONTACT
Deborah Swayne (dfs@research.att.com)
.SH SEE ALSO
Sample data files are included in the XGVis distribution.

For papers and further pointers, check the following web page:
.nf
http://www.research.att.com/areas/stat/xgobi/
.fi

.SH BUGS
Probably too numerous to mention, but you do the authors a favor if
you report the ones you notice.



Changes
September 2001:

  A great many extensions were added to xgvis, and
  a few bugs repaired in xgobi.

August 5 2000:

  Sphering is handled as a transformation instead of being
  buried inside the touring code.

January 3 2000:

  The initial CVS repository on franz, really just a developer's
  version.  The CORBA functionality is working, but still needs
  work, and is not yet documented.  There is a problem with the
  combination of Imakefile/GNUmakefile.

  The menus on each variable are now variable selection menus
  rather than transformation menus.

late November 1999:

  Variable transformations have been moved from the variable
  labels and given their own panel.  A variety of power transformations
  have been replaced by the Box-Cox family of transformations.  
  Transformation has been given a preparatory stage, too, which
  we're calling "domain adjustment:"  ie, ensure that the data
  you want to transform is within the domain of the transformation
  function.  (We're preparing to break transformation into two
  stages, with permutation, sorting and sphering moved to the
  second stage.)

  1D plotting:  David Scott's 1D ASH (Average Shifted Histogram)
  has been added to the textured dotplot as an alternative 1D plot.
  You can now display these plots either horizontally or vertically

  Axes:  The axes will now behave a bit better -- the labels should
  be smart enough not to draw on top of each other.  In addition,
  the Display menu has one more option, for displaying gridlines
  when in the 1D and 2D plots when axes are shown.

  Identify:  no label will be drawn unless you're within a few
  hundred pixels of some point.

  Linked brushing in the presence of both row groups and nlinkable: 
  few of you will have discovered these options, so even fewer of
  you will know that there were tremendous bugs and limitations
  if you wanted to use them together.  It was even worse if you used
  both of those options in conjunction with either subsetting or
  excluding points.  All this should now work.

  CORBA:  If you go so far as to read the source code, you'll notice
  references to CORBA.  This reflects work done by Duncan Temple
  Lang and Ross Ihaka to make xgobi available from omegahat
  (www.omegahat.org); we expect to test and release this work
  in a few weeks.

late June 1999:
  There are more revisions of the parallel coordinates display -- finally
  it can display labels instead of variable indices.  Long labels will
  be truncated to fit.  If you'd like to supply shorter labels, add them
  to the fname.col file like this:  longname|shortname
  
  Erase brushing has been completely redesigned.  There's no longer an
  erase operation on the brushing panel.  Instead, you brush the points
  you'd like to erase with some chosen glyph/color, and then open the
  "Hide or exclude" tool.  Using that tool, you can hide cases (ie,
  don't draw them) or exclude them (ie, don't draw them <and> don't
  include them in calculations for scaling or sphering or projection
  pursuit.)  If that seems too elaborate for a quick-and-dirty erase,
  just brush them in the background color.

early June 1999:
  There are some revisions in the parallel coordinates display --
  there are now actual parallel coordinates drawn.

  There's a new pair of command line arguments:
   -only n/N : read only a sample of n of N lines
   -only a,n : read only from line a to a+n

  The subset panel has been redesigned in the hope that it
  will be easier to use; its functionality has not changed.

  A couple of new transformations have been added; some of
  these will eventually migrate to a new panel entitled
  "Inference."

  The label of the currently identified point can be captured
  using the 'w' key, and then pasted into another window.

  xgvis changes:
  
  New command line arguments:
    -dims n   specify the dimensionality of the embedding space
    -stepsize x  specify the initial stepsize, a value between 0 and 1
  
March 1999:
  The biggest change in this distribution is the addition
  of the -scatmat command line argument, which instructs
  xgobi to display the input data as a scatterplot matrix,
  which can be linked to other xgobis.

This release (November 1998) has the
following changes:

 Probably the biggest change in this distribution is that
 another program, xgvis, has been included.  xgvis is an
 interactive multidimensional scaling program which has
 its own control panel but uses xgobi as its main data
 display.


This release (March 1998) has the following changes:

 XGobi supports Remote Procedure Calls (RPCs) as a means
 forinterprocess communication (IPC). For example, a link among
 XGobi, ArcView, and XploRe is currently being supported.  See
 http://www.galaxy.gmu.edu/~symanzik/AXX/ for more details.

 There's a new mode on the view menu -- moving points.  This
 is not intended as a way to fix outliers; rather, it's useful
 for graph viewing or manipulating maps or other structures
 that are drawn along with data.

 rgroups, lgroups:  a method for grouping rows so that brushing
 one member causes all other members of the same group to be
 brushed at the same time.  The method is extended to lines, too:
 all lines between points in the same group form a lines group.

This release (October, 1997) has the following changes:

  There is a new command line option, -subset n, that allows
  a subset of the data to be specified:  that is, all the data
  is read in, but an intial subset is used for display.

  The main control panel has been redesigned, so that it
  now uses several menus.  Everything on the View menu
  has been there for a while, while much of what's on the
  Tools menu is new:  Smoothing, jittering, the variable
  and case lists, and some methods for working with missing
  data.

  The case profile has been renamed 'parallel coordinates
  plot,' since that's what it is if all profiles are displayed.

  The file I/O has been gathered from here and there on

S/xgvis.S

xgvis <-
  function(dmat       = NULL,
           edges      = NULL,
           pos        = NULL,
           rowlab     = NULL,
           colors     = NULL,
           glyphs     = NULL,
           erase      = NULL,
           lines      = NULL,
           linecolors = NULL,
           resources  = NULL,
           display    = NULL)
{
  if (missing(edges) && missing(pos) && missing(dmat)) {
    cat("One of dmat, edges, or pos must be present\n")
    return()
  }

  tpath <- getenv("S_TMP")
  if(nchar(tpath) == 0) tpath <- getenv("TEMP")
  basefile <- paste(tpath, tempfile("xgvis"), sep = "/")

  ### distance matrix ###
  if (!missing(dmat)) {
    dmat <- eval(dmat)
    if (sum(abs(dmat[!is.na(dmat)])) == Inf > 0) {
      cat("xgvis can't handle Infs in dmat, replaced with NA\n")
      dmat[dmat==Inf] <- NA
    }
    dfile <- paste(basefile, ".dist", sep="")
    write(t(dmat), file = dfile, ncolumns = ncol(dmat))
    on.exit(unlink(dfile), add=T)
  }

  ### Edges ###
  if (!missing(edges))
    {
      # check data type
      if (!is.matrix(edges) || !is.numeric(edges) || dim(edges)[2] != 2) {
        cat("The 'edges' argument must be a numeric 2-column matrix\n")
        return()
      }

      edgesfile <- paste(basefile, ".edges", sep="")
      if (nrow(edges) > 0) {
	write(t(edges), file = edgesfile, ncol=2)
      }
      on.exit(unlink(edgesfile), add=T)
    }

  ### position matrix ###
  if (!missing(pos)) {
    pos <- eval(pos)
    if (sum(abs(pos[!is.na(pos)])) == Inf > 0)
      {
        cat("xgvis can't handle Inf in pos; replaced with NA\n")
        pos[pos==Inf] <- NA
      }
    pfile <- paste(basefile, ".pos", sep="")
    write(t(pos), file = pfile, ncolumns = ncol(pos))
    on.exit(unlink(pfile), add = T)
  }

  ### Row labels ###
  if (!missing(rowlab))
    # check data type
    if (!is.vector(rowlab) || !is.character(rowlab)) {
      cat("The 'rowlab' argument needs to be a character vector\n")
      return()
    }
  if (length(rowlab) > 0) {
    rowfile <- paste(basefile, ".row", sep="")
    write(rowlab, file = rowfile, ncol=1)
    on.exit(unlink(rowfile), add = T)
  }

  ### Colors ###
  if (!missing(colors)) {
    # check data type
    if (!is.vector(colors) || !is.character(colors)) {
      cat("The 'colors' argument needs to be a character vector\n")
      return()
    }
    colorfile <- paste(basefile, ".colors", sep="")
    write(colors, file = colorfile, ncol=1)
    on.exit(unlink(colorfile), add = T)
  }
  
  ### Glyphs ###
  if (!missing(glyphs)) {
    # check data type
    if (!is.vector(glyphs) || !is.numeric(glyphs)) {
      cat("The 'glyphs' argument needs to be a numeric vector\n")
      return()
    }
    glyphfile <- paste(basefile, ".glyphs", sep="")
    write(glyphs, file = glyphfile, ncol=1)
    on.exit(unlink(glyphfile), add = T)
  }
  
  ### Erase ###
  if (!missing(erase)) {
    # check data type
    if (!is.vector(erase) || !is.numeric(erase)) {
      cat("The 'erase' argument needs to be a numeric vector\n")
      return()
    }
    erasefile <- paste(basefile, ".erase", sep="")
    write(erase, file = erasefile, ncol=1)
    on.exit(unlink(erasefile), add = T)
  }

  ### Connected lines ###
  if (!missing(lines)) {
    # check data type
    if (!is.matrix(lines) || !is.numeric(lines) || dim(lines)[2] != 2) {
      cat("The 'lines' argument must be a numeric 2-column matrix\n")
      return()
    }

    linesfile <- paste(basefile, ".lines", sep="")
    if (nrow(lines) > 0) {
      write(t(lines), file = linesfile, ncol=2)
      on.exit(unlink(linesfile), add = T)
    }
  }

  ### Line colors ###
  if ((!missing(lines) || !missing(edges)) && !missing(linecolors)) {
    # check data type
    if (!is.vector(linecolors) || !is.character(linecolors)) {
      cat("The 'linecolors' argument must be a character vector\n")
      return()
    }
    linecolorfile <- paste(basefile, ".linecolors", sep="")
    write(linecolors, file = linecolorfile, ncol=1)
    on.exit(unlink(linecolorfile), add = T)
  }

  ### Resources ###
  if (!missing(resources)) {
    # check data type
    if (!is.vector(resources) || !is.character(resources)) {
      cat("The 'resources' argument must be a character vector\n")
      return()
    }
    resourcefile <- paste(basefile, ".resources", sep="")
    write(resources, file = resourcefile, ncol=1)
    on.exit(unlink(resourcefile), add = T)
  }


  # Note to installer:
  # Here you need to specify the path to the xgvis batch file/ executable
  # on your system.
  #
  # dos example:
#  xgpath <- "c:/packages/xgobi/xgvis.bat"
  # unix example:
  xgpath <- "/usr/dfs/xgobi/joint/src/xgvis"
  command <- paste(xgpath, basefile)
  cat(command, "\n")
  # dos:
#  invisible(dos(command, multi = F, minimized = T, output.to.S = F, translate = T))
  # unix:
  invisible(unix(command))
}

